### DISCUSSION TITLE
Unified primary UV Map read/write behavior

### DISCUSSION THREAD

**Discussion by BSalem** [Autodesk/maya-usd#1477] [General] [answered]
Hi all,
I suggest that we keep only one behavior for writing the primary UV map and not to put an option for doing something else!
By giving an option to write map1 as st or as map1, we are opening a door for mistakes and the folks will end up by producing assets that can render properly here and not there... By consider that USD is and should always be (Universal), then we shouldn't open a door for something like this.

The current default behavior that the primary UV goes as map1, means that you cannot properly render these USDs with renderman (the founder) before edit them and making another copy! or adding a node that copies from primvars:map1 to primvars:st which is considered a performance issue.

So, I vote here that we stick with Pixar's defaults, that the primary UV map should go as 'st' just as how it used to be since day one, then any development can consider that this is how the USD look like (the primary UV map is called 'st'), this way, both old development and old USD assets will keep working as expected, while also on the other hand, new development can follow this rule easily (Maya won't fail to deal with a UV set no matter what its name - but renderman will).

Note also that in many cases, people may mix render engines, and use different applications/renderers for the same shot, it's rare but it happens, so, this is another reason why I vote for Keeping the USDs universal and usable as expected in everywhere with every renderer.

**Comment by kxl-adsk** [comment, marked as answer]
Hi @BSalem,

One would argue that hardcoding a name is not something very universal. Having said that we have been contacted by customers regarding the situation with UV import/export and did some investigation. We will be soon proposing a change that will rework PR876 and:
- We will bring back legacy behavior for index 0 and remove `MAYAUSD_EXPORT_MAP1_AS_PRIMARY_UV_SET` and `MAYAUSD_IMPORT_PRIMARY_UV_SET_AS_MAP1`
- We will append index to `st` the primvar when more than one UV stream is available, e.g. `st`, `st1`, `st2`, `st3`
- We will support roundtrip of UV names with custom metadata attached to the primvar
- We will preserve the indexing on import
- We will generate materials override only when necessary
- Files exported with the previous default mode will continue to properly import. On export, we will perform actions as described above.

This should address your concerns and ones raised in this forum by @koen-v.

Cheers,
KxL

**Comment by BSalem** [comment]
super awesome...
I totally agree with not hardcoding, I just meant that we get to something compatible with everything, the thing that you've already considered in this implementation.

Thanks

**Reply by kxl-adsk** [reply]
FYI. https://github.com/Autodesk/maya-usd/pull/1528
